File: backend/src/models/ticket.py
from beanie import Document,Link, PydanticObjectId
from pydantic import Field
from typing import Optional, List, Dict
from datetime import datetime, timezone
from .category import Category
from .subcategory import SubCategory
from .enums import TicketStatus, TicketPriority
from .user import User
from .rich_text import RichTextContent
class Ticket(Document):
    id: Optional[PydanticObjectId] = Field(default=None, alias="_id", description="Primary key (MongoDB ObjectId)")
    category_id: Link[Category] = Field(..., description="ID of the category this ticket belongs to", alias="categoryId")
    sub_category_id: Link[SubCategory] = Field(..., description="ID of the subcategory this ticket belongs to", alias="subCategoryId")
    user_id: Link[User] = Field(..., description="ID of the user who created the ticket", alias="userId")
    agent_id: Optional[Link[User]] = Field(None, description="ID of the agent assigned to the ticket (if any)", alias="agentId")
    
    title: str = Field(..., description="Title of the ticket")
    description: str = Field(..., description="Detailed description of the ticket")
    content: RichTextContent = Field(..., description="Rich text content of the ticket with HTML, JSON, and plain text formats")
    tag_ids: Optional[List[Dict[str, str]]] = Field(default_factory=list, description="List of tag IDs associated with the ticket", alias="tagIds")

    status: TicketStatus = Field(default=TicketStatus.new, description="Status of the ticket") 
    priority: TicketPriority = Field(default=TicketPriority.medium, description="Priority: low, medium, high, critical")

    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), alias="createdAt")
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc), alias="updatedAt")
    closed_at: Optional[datetime] = Field(None, alias="closedAt")

    class Settings:
        name = "tickets"  # MongoDB collection name

    

File: frontend/src/app/shared/types/ticket.ts
// Ticket-related types matching backend API
import { Category } from './category';
import { SubCategory } from './subcategory';

export type TicketStatus = 'new' | 'in_progress' | 'waiting_for_customer' | 'waiting_for_agent' | 'resolved' | 'closed';
export type TicketPriority = 'low' | 'medium' | 'high' | 'critical';

// Rich text content structure
export interface RichTextContent {
  html: string;
  json: Record<string, unknown>;
  text: string;
}

export interface Ticket {
  id: string;
  title: string;
  description: string;
  content: RichTextContent;
  category: Category;
  subCategory: SubCategory;
  userInfo: UserInfo;
  agentInfo?: UserInfo;
  status: TicketStatus;
  priority: TicketPriority;
  createdAt: string;
  updatedAt: string;
  closedAt?: string;
  // ENHANCEMENT L1 AI TICKET SUMMARY - Add AI summary fields
  aiSummary?: string;
  summaryGeneratedAt?: string;
  // ENHANCEMENT L2 SLA AUTOMATION - Add SLA fields
  slaDueDate?: string;
  slaBreached?: boolean;
  slaPausedAt?: string;
  slaTotalPausedTime?: number;
}

export interface UserInfo {
  id: string;
  email: string;
  name?: string;
  role?: string; // "user" or "agent" - helps identify user type in comments and tickets
}

export interface CreateTicketRequest {
  category_id: string;
  sub_category_id: string;
  title: string;
  description: string;
  content: RichTextContent;
  priority?: TicketPriority;
}

export interface UpdateTicketRequest {
  title?: string;
  description?: string;
  content?: RichTextContent;
  status?: TicketStatus;
  priority?: TicketPriority;
  agent_id?: string;
}

export interface TicketStats {
  total: number;
  new: number;
  in_progress: number;
  waiting_for_customer: number;
  waiting_for_agent: number;
  resolved: number;
  closed: number;
  by_priority: {
    low: number;
    medium: number;
    high: number;
    critical: number;
  };
}

export interface Comment {
  id: string;
  ticket_id: string;
  user: UserInfo;
  content: RichTextContent;
  createdAt: string; // Changed from created_at to match backend alias
  updatedAt: string; // Changed from updated_at to match backend alias
  // ENHANCEMENT L1 COMMENT EDITING - Add edit tracking fields
  edited?: boolean;
  edit_count?: number;
  edit_history?: CommentEditHistory[];
}

// ENHANCEMENT L1 COMMENT EDITING - Edit history tracking
export interface CommentEditHistory {
  edited_at: string;
  previous_content?: RichTextContent;
}

export interface CreateComment {
  content: RichTextContent;
}

export interface UpdateComment {
  content: RichTextContent;
}

// ENHANCEMENT L1 AI TICKET SUMMARY - AI summary types
export interface TicketSummaryResponse {
  summary: string;
}

export interface ClosingCommentsResponse {
  suggestions: string[];
}

export type CreateTicket = CreateTicketRequest;
export type UpdateTicket = UpdateTicketRequest;

export interface TicketAssignment {
  agent_id: string;
}

export interface TicketClosure {
  resolution: string;
}

// ENHANCEMENT L1 AI CLOSING SUGGESTIONS - AI closing suggestion types
export interface ClosingCommentsResponse {
  reason: string;
  comment: string;
}